src/NServiceBus.Core/Transports/Msmq/MsmqFastHeadersParser.cs

```diff
@@ -0,0 +1,239 @@
+            }
+        }
+
+        MsmqFastHeadersParser(byte[] bytes)
+        {
+            this.bytes = bytes;
+        }
+
+        public static Dictionary<string, string> ParseHeaders(byte[] bytes)
+        {
+            return new MsmqFastHeadersParser(bytes).ParseHeaders();
+        }
+
+        Dictionary<string, string> ParseHeaders()
+        {
+            var result = new Dictionary<string, string>();
+            var byteCount = bytes.Length;
+
+            if (byteCount == 0)
+            {
+                return result;
+            }
+
+            // <xml
+            if (TryMoveToNext(Open) == false)
+            {
+                return result;
+            }
+            pos += 1;
+
+            // <ArrayOfHeaderInfo
+            if (TryMoveToNext(Open) == false)
+            {
+                return result;
+            }
+            pos += 1;
+
+            // loop over <HeaderInfo>, moving to the starting tag
+            while (TryMoveToNext(Open))
+            {
+                if (TryConsumeTag(HeaderInfo) == false)
+                {
+                    // this is not a header info
+                    break;
+                }
+                // inside <HeaderInfo>...
+                //                    ^
+
+                ConsumeTag(HeaderKey);
+                // <Key>...
+                //      ^
+                var keyStart = pos;
+                MoveToNext(Open);
+
+                var key = GetKey(new ArraySegment<byte>(bytes, keyStart, pos - keyStart));
+
+                ConsumeTag(HeaderKeyEnd);
+                // </Key>...
+                //       ^
+
+                MoveToNext(Open);
+
+                if (TryConsumeTag(HeaderValue))
+                {
+                    // <Value>...
+                    //        ^
+
+                    var valueStart = pos;
+                    MoveToNext(Open);
+
+                    var value = GetValue(new ArraySegment<byte>(bytes, valueStart, pos - valueStart));
+                    result.Add(key, value);
+
+                    TryConsumeTag(HeaderValueEnd);
+                    // </Value>...
+                    //         ^
+                }
+                else
+                {
+                    if (TryConsumeTag(HeaderValueEmpty))
+                    {
+                        // <Value />...
+                        //          ^
+                        result.Add(key, "");
+                    }
+                    else
+                    {
+                        throw new SerializationException();
+                    }
+                }
+                ConsumeTag(HeaderInfoEnd);
+                // </HeaderInfo>...
+                //              ^
+            }
+
+            return result;
+        }
+
+        static string GetKey(ArraySegment<byte> s)
+        {
+            string key;
+            return KnownHeaders.TryGetValue(s, out key) ? key : Encoding.UTF8.GetString(s.Array, s.Offset, s.Count);
```

Comment:
In the case where the header isn't found, I wonder if it would be worth trying to add it to the collection? If the transport defines any of its own header keys, then these would be missed by the current `KnownHeaders` initialization.